internal/factsengine/factscache/cache_test.go

```diff
@@ -0,0 +1,173 @@
+	suite.Suite
+	returnValue string
+	count       int
+}
+
+func TestFactsCacheTestSuite(t *testing.T) {
+	suite.Run(t, new(FactsCacheTestSuite))
+}
+
+func (suite *FactsCacheTestSuite) SetupSuite() {
+	suite.returnValue = "value"
+}
+
+func (suite *FactsCacheTestSuite) SetupTest() {
+	suite.count = 0
+}
+
+// nolint:errcheck
+func (suite *FactsCacheTestSuite) TestEntries() {
+	cache := factscache.NewFactsCache()
+	cache.GetOrUpdate("entry1", func(args ...interface{}) (interface{}, error) {
+		return "", nil
+	})
+	cache.GetOrUpdate("entry2", func(args ...interface{}) (interface{}, error) {
+		return "", nil
+	})
+	entries := cache.Entries()
+
+	suite.ElementsMatch([]string{"entry1", "entry2"}, entries)
+}
+
+func (suite *FactsCacheTestSuite) TestGetOrUpdate() {
+	cache := factscache.NewFactsCache()
+
+	updateFunc := func(args ...interface{}) (interface{}, error) {
+		return suite.returnValue, nil
+	}
+
+	value, err := cache.GetOrUpdate("entry1", updateFunc)
+
+	suite.Equal(suite.returnValue, value)
+	suite.NoError(err)
+}
+
+func (suite *FactsCacheTestSuite) TestGetOrUpdateWithError() {
+	cache := factscache.NewFactsCache()
+	someError := "some error"
+
+	updateFunc := func(args ...interface{}) (interface{}, error) {
+		return nil, fmt.Errorf(someError)
+	}
+
+	_, err := cache.GetOrUpdate("entry", updateFunc)
+
+	suite.EqualError(err, someError)
+}
+
+func (suite *FactsCacheTestSuite) TestGetOrUpdateCacheHit() {
+	cache := factscache.NewFactsCache()
+
+	updateFunc := func(args ...interface{}) (interface{}, error) {
+		suite.count++
+		return suite.returnValue, nil
+	}
+
+	// nolint:errcheck
+	cache.GetOrUpdate("entry", updateFunc)
+	value, err := cache.GetOrUpdate("entry", updateFunc)
+
+	suite.Equal(suite.returnValue, value)
+	suite.Equal(1, suite.count)
+	suite.NoError(err)
+}
+
+func (suite *FactsCacheTestSuite) TestGetOrUpdateWithArgs() {
+	cache := factscache.NewFactsCache()
+
+	// nolint:forcetypeassert
+	updateFunc := func(args ...interface{}) (interface{}, error) {
+		arg1 := args[0].(int)
+		arg2 := args[1].(string)
+		return fmt.Sprintf("%d_%s", arg1, arg2), nil
+	}
+
+	value, err := cache.GetOrUpdate("entry", updateFunc, 1, "text")
+
+	suite.Equal("1_text", value)
+	suite.NoError(err)
+}
+
+// nolint:errcheck
+func (suite *FactsCacheTestSuite) TestGetOrUpdateCacheConcurrent() {
+	cache := factscache.NewFactsCache()
+	g := errgroup.Group{}
+
+	updateFunc := func(args ...interface{}) (interface{}, error) {
+		value, _ := args[0].(string)
+		time.Sleep(100 * time.Millisecond)
+		return value, nil
+	}
+
+	g.Go(func() error {
+		value, _ := cache.GetOrUpdate("entry1", updateFunc, "value1")
+		castedValue, _ := value.(string)
+		suite.Equal("value1", castedValue)
+		return nil
+	})
+	g.Go(func() error {
+		value, _ := cache.GetOrUpdate("entry2", updateFunc, "value3")
+		castedValue, _ := value.(string)
+		suite.Equal("value3", castedValue)
+		return nil
+	})
+	time.Sleep(50 * time.Millisecond)
+	// The next 2 calls return the memoized value
```

Comment:
Maybe it's a bit overkill, but just wondering if we should we test that these 2 following calls are not blocking for 100ms and instead return immediately?

Reply:
how can you test that? I mean, the time that the test needs. Is that even possible?
Besides, they don't return immediately, the get blocked until the first one is finished, and both return at the exact same moment with the value from the first execution.